File: javascript/src/transport.ts
import {
  AuthenticationError,
  NetworkError,
  NotFoundError,
  QuotaExceededError,
  RateLimitError,
  ServerError,
  TierAccessError,
  ValidationError,
} from "./errors";
import type { QuotaInfo } from "./models/common";

export interface RequestOptions {
  method: string;
  path: string;
  body?: unknown;
  params?: Record<string, string | number | undefined>;
  headers?: Record<string, string>;
  signal?: AbortSignal;
}

export interface TransportConfig {
  apiKey: string;
  baseUrl: string;
  timeout: number;
  maxRetries: number;
  customHeaders: Record<string, string>;
}

export class Transport {
  private config: TransportConfig;
  quota: QuotaInfo = {};

  constructor(config: TransportConfig) {
    this.config = config;
  }

  async request<T>(opts: RequestOptions): Promise<T> {
    let lastError: Error | undefined;

    for (let attempt = 0; attempt <= this.config.maxRetries; attempt++) {
      try {
        const url = this.buildUrl(opts.path, opts.params);
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.config.timeout);

        const headers: Record<string, string> = {
          "X-API-Key": this.config.apiKey,
          "Content-Type": "application/json",
          Accept: "application/json",
          "User-Agent": "aegis-js-sdk/0.1.0",
          ...this.config.customHeaders,
          ...opts.headers,
        };

        const response = await fetch(url, {
          method: opts.method,
          headers,
          body: opts.body ? JSON.stringify(opts.body) : undefined,
          signal: opts.signal ?? controller.signal,
        });
        clearTimeout(timeoutId);

        const requestId = response.headers.get("x-request-id") ?? undefined;
        this.updateQuota(response.headers);

        if (response.status === 429 && attempt < this.config.maxRetries) {
          const retryAfter = response.headers.get("retry-after");
          const wait = retryAfter ? parseFloat(retryAfter) * 1000 : this.backoff(attempt);
          await sleep(wait);
          continue;
        }

        if (response.status >= 500 && attempt < this.config.maxRetries) {
          await sleep(this.backoff(attempt));
          continue;
        }

        if (!response.ok) {
          const body = await response.json().catch(() => ({ error: response.statusText }));
          this.throwForStatus(response.status, body, requestId);
        }

        if (response.status === 204) return undefined as T;
        return (await response.json()) as T;
      } catch (err) {
        if (err instanceof AuthenticationError || err instanceof TierAccessError ||
            err instanceof ValidationError || err instanceof NotFoundError ||
            err instanceof QuotaExceededError) {
          throw err;
        }
        lastError = err as Error;
        if (attempt < this.config.maxRetries) {
          await sleep(this.backoff(attempt));
          continue;
        }
      }
    }
    throw new NetworkError("Max retries exceeded", lastError);
  }

  async *stream(opts: RequestOptions): AsyncGenerator<string> {
    const url = this.buildUrl(opts.path, opts.params);
    const headers: Record<string, string> = {
      "X-API-Key": this.config.apiKey,
      "Content-Type": "application/json",
      Accept: "text/event-stream",
      "User-Agent": "aegis-js-sdk/0.1.0",
      ...this.config.customHeaders,
    };

    const response = await fetch(url, {
      method: opts.method,
      headers,
      body: opts.body ? JSON.stringify(opts.body) : undefined,
    });

    if (!response.ok) {
      const body = await response.json().catch(() => ({ error: response.statusText }));
      const requestId = response.headers.get("x-request-id") ?? undefined;
      this.throwForStatus(response.status, body, requestId);
    }

    if (!response.body) return;
    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split("\n");
        buffer = lines.pop() ?? "";

        for (const line of lines) {
          if (line.startsWith("data:")) {
            const data = line.slice(5).trim();
            if (data === "[DONE]") return;
            yield data;
          }
        }
      }
    } finally {
      reader.releaseLock();
    }
  }

  private buildUrl(path: string, params?: Record<string, string | number | undefined>): string {
    const url = new URL(path, this.config.baseUrl);
    if (params) {
      for (const [key, value] of Object.entries(params)) {
        if (value !== undefined) url.searchParams.set(key, String(value));
      }
    }
    return url.toString();
  }

  private updateQuota(headers: Headers): void {
    const limit = headers.get("x-quota-limit");
    const used = headers.get("x-quota-used");
    const remaining = headers.get("x-quota-remaining");
    if (limit) this.quota.limit = parseInt(limit, 10);
    if (used) this.quota.used = parseInt(used, 10);
    if (remaining) this.quota.remaining = parseInt(remaining, 10);
  }

  private throwForStatus(
    status: number,
    body: Record<string, unknown>,
    requestId?: string,
  ): never {
    const message = (body.error ?? body.message ?? `HTTP ${status}`) as string;
    const opts = { requestId, body };

    if (status === 401) throw new AuthenticationError(message, opts);
    if (status === 403) throw new TierAccessError(message, { ...opts, requiredTier: body.required_tier as string, upgradeUrl: body.upgrade_url as string });
    if (status === 404) throw new NotFoundError(message, opts);
    if (status === 429) {
      if (body.code === "QUOTA_EXCEEDED" || (message.toLowerCase().includes("quota"))) {
        throw new QuotaExceededError(message, { ...opts, limit: body.limit as number, used: body.used as number, resetAt: body.reset_at as string });
      }
      throw new RateLimitError(message, opts);
    }
    if (status === 400 || status === 422) throw new ValidationError(message, status, opts);
    if (status >= 500) throw new ServerError(message, status, opts);
    throw new ServerError(message, status, opts);
  }

  private backoff(attempt: number): number {
    return 500 * Math.pow(2, attempt);
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: rust/src/transport.rs
use std::sync::RwLock;
use std::time::Duration;

use reqwest::{header, Client, Method, StatusCode};
use serde::de::DeserializeOwned;
use serde::Serialize;

use crate::config::ClientConfig;
use crate::error::{AegisError, ApiErrorBody, Result};
use crate::models::common::QuotaInfo;

const SDK_VERSION: &str = "0.1.0";
const USER_AGENT: &str = "aegis-rust-sdk/0.1.0";

pub(crate) struct Transport {
    client: Client,
    config: ClientConfig,
    last_quota: RwLock<Option<QuotaInfo>>,
}

impl Transport {
    pub fn new(config: ClientConfig) -> Result<Self> {
        let mut default_headers = header::HeaderMap::new();
        default_headers.insert(
            "X-API-Key",
            header::HeaderValue::from_str(&config.api_key)
                .map_err(|_| AegisError::Authentication {
                    message: "Invalid API key format".into(),
                    request_id: None,
                })?,
        );
        default_headers.insert(
            header::USER_AGENT,
            header::HeaderValue::from_static(USER_AGENT),
        );
        default_headers.insert(
            "X-SDK-Version",
            header::HeaderValue::from_static(SDK_VERSION),
        );

        for (key, value) in &config.custom_headers {
            if let (Ok(name), Ok(val)) = (
                header::HeaderName::from_bytes(key.as_bytes()),
                header::HeaderValue::from_str(value),
            ) {
                default_headers.insert(name, val);
            }
        }

        let client = Client::builder()
            .timeout(config.timeout)
            .default_headers(default_headers)
            .build()?;

        Ok(Self {
            client,
            config,
            last_quota: RwLock::new(None),
        })
    }

    pub async fn request<T: DeserializeOwned>(
        &self,
        method: Method,
        path: &str,
        body: Option<&impl Serialize>,
    ) -> Result<T> {
        let response = self.request_with_retry(method, path, body).await?;
        self.extract_quota(&response);
        let bytes = response.bytes().await?;
        serde_json::from_slice(&bytes).map_err(AegisError::from)
    }

    pub async fn request_raw(
        &self,
        method: Method,
        path: &str,
        body: Option<&impl Serialize>,
    ) -> Result<reqwest::Response> {
        let response = self.request_with_retry(method, path, body).await?;
        self.extract_quota(&response);
        Ok(response)
    }

    pub fn last_quota(&self) -> Option<QuotaInfo> {
        self.last_quota.read().ok().and_then(|q| q.clone())
    }

    async fn request_with_retry(
        &self,
        method: Method,
        path: &str,
        body: Option<&impl Serialize>,
    ) -> Result<reqwest::Response> {
        let url = format!("{}{}", self.config.base_url, path);
        let mut last_err: Option<AegisError> = None;
        let max_attempts = self.config.max_retries + 1;

        for attempt in 0..max_attempts {
            if attempt > 0 {
                let backoff = Duration::from_millis(100 * 2u64.pow(attempt - 1));
                tokio::time::sleep(backoff).await;
            }

            let mut builder = self.client.request(method.clone(), &url);
            if let Some(b) = body {
                builder = builder.json(b);
            }

            let response = match builder.send().await {
                Ok(r) => r,
                Err(e) => {
                    last_err = Some(AegisError::Network(e));
                    continue;
                }
            };

            let status = response.status();

            if status.is_success() {
                return Ok(response);
            }

            if Self::is_retryable(status) && attempt < max_attempts - 1 {
                if status == StatusCode::TOO_MANY_REQUESTS {
                    if let Some(retry_after) = response
                        .headers()
                        .get("retry-after")
                        .and_then(|v| v.to_str().ok())
                        .and_then(|v| v.parse::<f64>().ok())
                    {
                        tokio::time::sleep(Duration::from_secs_f64(retry_after)).await;
                    }
                }
                let body_bytes = response.bytes().await.unwrap_or_default();
                let err_body: ApiErrorBody =
                    serde_json::from_slice(&body_bytes).unwrap_or(ApiErrorBody {
                        message: Some(format!("HTTP {}", status.as_u16())),
                        error: None,
                        request_id: None,
                        required_tier: None,
                        upgrade_url: None,
                        limit: None,
                        used: None,
                        reset_at: None,
                        retry_after: None,
                    });
                last_err = Some(Self::map_status_error(status, err_body));
                continue;
            }

            let body_bytes = response.bytes().await.unwrap_or_default();
            let err_body: ApiErrorBody =
                serde_json::from_slice(&body_bytes).unwrap_or(ApiErrorBody {
                    message: Some(format!("HTTP {}", status.as_u16())),
                    error: None,
                    request_id: None,
                    required_tier: None,
                    upgrade_url: None,
                    limit: None,
                    used: None,
                    reset_at: None,
                    retry_after: None,
                });
            return Err(Self::map_status_error(status, err_body));
        }

        Err(last_err.unwrap_or(AegisError::Server {
            message: "Max retries exceeded".into(),
            status_code: 0,
        }))
    }

    fn is_retryable(status: StatusCode) -> bool {
        status == StatusCode::TOO_MANY_REQUESTS
            || status.is_server_error()
    }

    fn map_status_error(status: StatusCode, body: ApiErrorBody) -> AegisError {
        match status {
            StatusCode::UNAUTHORIZED => AegisError::Authentication {
                message: body.message(),
                request_id: body.request_id,
            },
            StatusCode::FORBIDDEN => AegisError::TierAccess {
                message: body.message(),
                required_tier: body.required_tier,
                upgrade_url: body.upgrade_url,
            },
            StatusCode::NOT_FOUND => AegisError::NotFound {
                message: body.message(),
            },
            StatusCode::TOO_MANY_REQUESTS => {
                if body.limit.is_some() || body.used.is_some() {
                    AegisError::QuotaExceeded {
                        message: body.message(),
                        limit: body.limit,
                        used: body.used,
                        reset_at: body.reset_at,
                    }
                } else {
                    AegisError::RateLimit {
                        message: body.message(),
                        retry_after: body.retry_after,
                    }
                }
            }
            s if s == StatusCode::BAD_REQUEST || s == StatusCode::UNPROCESSABLE_ENTITY => {
                AegisError::Validation {
                    message: body.message(),
                    status_code: s.as_u16(),
                }
            }
            s => AegisError::Server {
                message: body.message(),
                status_code: s.as_u16(),
            },
        }
    }

    fn extract_quota(&self, response: &reqwest::Response) {
        let headers = response.headers();
        let get_header = |name: &str| -> Option<String> {
            headers.get(name).and_then(|v| v.to_str().ok()).map(String::from)
        };
        let get_i64 = |name: &str| -> Option<i64> {
            headers.get(name).and_then(|v| v.to_str().ok()).and_then(|v| v.parse().ok())
        };

        let limit = get_i64("X-Quota-Limit");
        let used = get_i64("X-Quota-Used");
        let remaining = get_i64("X-Quota-Remaining");
        let reset_at = get_header("X-Quota-Reset");

        if limit.is_some() || used.is_some() || remaining.is_some() {
            let info = QuotaInfo {
                limit,
                used,
                remaining,
                reset_at,
            };
            if let Ok(mut guard) = self.last_quota.write() {
                *guard = Some(info);
            }
        }
    }
}
